MPI_Reduce: Bad NULL buffer checks for intercommunicators

When used with intercommunicators, the `MPI_Reduce` routine should ignore `sendbuf` if `root in {MPI_ROOT, MPI_PROC_NULL}` and should ignore `recvbuf` if `root >= 0` (See MPI-4 document, pp. 266 lines 16-18). MPICH seems to be performing some NULL checks that IMHO are wrong. 

Look at the following compute-pi reproducer:
```
git clone https://github.com/mpi4py/mpi4py.git
$ make LANGS=c -C mpi4py/demo/spawning/
make: Entering directory '/home/dalcinl/Devel/mpi4py/demo/spawning'
./cpi-master-c.exe -> ./cpi-worker-c.exe
Fatal error in internal_Reduce: Invalid buffer pointer, error stack:
internal_Reduce(150): MPI_Reduce(sendbuf=(nil), recvbuf=0x7ffd23a8dcb0, count=1, MPI_DOUBLE, MPI_SUM, MPI_ROOT, comm=0x84000000) failed
internal_Reduce(114): Null buffer pointer
make: *** [makefile:14: test] Error 1
make: Leaving directory '/home/dalcinl/Devel/mpi4py/demo/spawning'
```

I have not checked all the other variants (non blocking, persistent, large count) of `MPI_Reduce`, but they may have the same issue. This is the list of all routines to check:
* [x] `MPI_Reduce`
* [ ] `MPI_Ireduce`
* [ ] `MPI_Reduce_init`
* [ ] `MPI_Reduce_c`
* [ ] `MPI_Ireduce_c`
* [ ] `MPI_Reduce_init_c`

--- a/maint/local_python/binding_c.py
+++ b/maint/local_python/binding_c.py
@@ -2175,12 +2175,16 @@ def dump_validate_userbuffer_reduce(func, sbuf, rbuf, ct, dt, op):
         G.out.append("    }")
         G.out.append("}")
         # test sendbuf
+        cond_a = cond_intra
+        cond_b = cond_inter + " && root != MPI_ROOT && root != MPI_PROC_NULL"
+        dump_if_open("(" + cond_a + ") || (" + cond_b + ")")
         G.out.append("if (" + cond_inter + ") {")
         G.out.append("    MPIR_ERRTEST_SENDBUF_INPLACE(%s, %s, mpi_errno);" % (sbuf, ct))
         G.out.append("}")
         G.out.append("if (%s > 0 && %s != MPI_IN_PLACE) {" % (ct, sbuf))
         G.out.append("    MPIR_ERRTEST_USERBUFFER(%s, %s, %s, mpi_errno);" % (sbuf, ct, dt))
         G.out.append("}")
+        dump_if_close()
 
         G.out.append("DEDENT")
         G.out.append("}")

